### PR TITLE
Fix/constant tensors

### LINKED ISSUE #461: Refactor and Simplify Device Handling
### Description
In the process of developing a Tensor module, it became clear that the current handling of device migration for modules can be improved and simplified. Specifically, the interaction between `to_device`, `map`, and `visit` methods is not intuitive and leads to confusion when trying to change the device of a module. 

Furthermore, we also found a problem with how the constant tensors are handled. It was observed that when the parent module is moved to a different device, the tensor field does not move to the new device. This needs to be addressed.

The code has evolved over time and seems to have been adapted to new requirements without a significant refactor. The team has recognized that there is a need for this code to be reviewed and restructured in order to simplify the mental model and minimize surprises for future developers.

### Steps

1. Refactor the device handling in the modules to improve and simplify the interaction between `to_device`, `map`, and `visit` methods.
2. Resolve the issue where the constant tensors do not move along with their parent module when changing devices.
3. Review and adjust the usage of the "B: Backend" reference in order to reduce confusion and align with the Rust standard lib behavior as closely as possible.
4. Perform an overall code review and refactor to simplify and clarify the codebase, keeping in mind the principle of adhering as closely as possible to the behaviors of "native" traits and derives in the Rust standard lib.

### Related Tickets

* https://github.com/burn-rs/burn/pull/459
* https://github.com/burn-rs/burn/issues/452

### PR DESCRIPTION
**Pull request by nathanielsimard**
In this effort to support constant tensors of any kind (`Int` and `Bool`, not just `Float`), I had to fix how we handle devices within modules.

Fix #461 #665 #982

### CODE REVIEW

File: burn-core/src/module/param/tensor.rs [hunk 1]
```diff
@@ -4,22 +4,36 @@ use crate::tensor::{
...
     Tensor,
 };
+use alloc::vec::Vec;
+use burn_tensor::{Bool, Int};
 
 impl<B: Backend, const D: usize> From<Tensor<B, D>> for Param<Tensor<B, D>> {
     fn from(value: Tensor<B, D>) -> Self {
         Param::new(ParamId::new(), value.require_grad())
     }
 }
```
Comment (antimora) [inline burn-core/src/module/param/tensor.rs:15]: Do we need to explain why we do not require grad?

**Review by antimora — APPROVED** [review]
Looks good. I had one comment. Please see below.

File: burn-core/src/module/param/tensor.rs (hunk 1 above)
Comment (nathanielsimard) [reply burn-core/src/module/param/tensor.rs:17]: I added why we add require grad, other tensor (bool and int) can't have gradients

**Review by dcvz — APPROVED** [review]
Nice work! 🔥 🚀 

The `devices()` needing devices looks a bit odd to me. Especially when it sometimes leads being used as `self.devices(Vec::new())` but I think I'm maybe missing something.

**Comment by antimora** [comment]
OK. We got a couple approvals and I need this PR for my onnx changes, hence I am merging.

**Comment by nathanielsimard** [comment]
> Nice work! 🔥 🚀
> 
> The `devices()` needing devices looks a bit odd to me. Especially when it sometimes leads being used as `self.devices(Vec::new())` but I think I'm maybe missing something.

Maybe we could rename the method, but we collect all the devices in the module tree in the provided vector. The alternative would imply a lot more allocations. Maybe we could rename the method to `collect_devices` and add a default implementation `devices` that uses `collect_devices` underneath.

**Pull request merged**